core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -105,16 +105,31 @@ class ReplicaFetcherThread(name:String,
       leaderEndOffset
     } else {
       /**
-       * The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
+       * If the leader's log end offset is greater than follower's log end offset, there are two possibilities:
+       * 1. The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
        * start offset because the leader has deleted old logs (log.logEndOffset < leaderStartOffset).
+       * 2. When leader migration occurs, there is a slight chance that the old leader's high watermark is greater than
+       * the new leader's log end offset, so an OffsetOutOfRangeException was thrown. After that some more messages are
+       * produced to the leader. When the follower was trying to handle the exception and query the log end offset of
+       * broker, the leader's log end offset become higher than follower's log end offset.
+       *
+       * In the first case, the follower's current log end offset is smaller than leader's log start offset. So the
+       * follower should truncate all its log, roll out a new segment and start to fetch from current leader's log
+       * start offset.
+       * In the second case, the follower should just keep the current log segments and retry fetch.
```

Comment:
Actually we can achieve what you said by apply appropriate configuration to broker and producer. i.e. consistency guarantee is handled at another level other than replica fetching exception handling.

I think the data mismatch might be unavoidable from the design when unclean leader election occurs. What if in step 4 the LEO on broker 1 has already grow beyond 100. There will be no OffsetOutOfRangeException thrown and the data would be different. Also, even if we truncate the data to on broker 0 to 90 as you said, because the  the high watermark has already been increased to 100, consumer might have already seen 90-100 on broker 0. Now if they reconsume from 90-100 on the new leader(broker 1), the data would still be different from what they saw before when they consumer from (broker 0).

I agree it is better to truncate the data to 90 if possible, but that seems a much bigger change that needs the leader to persist all the LEO of all followers. And we have to deal with cases such as if a broker left a group should we remove its LEO or not.